### DISCUSSION TITLE
Schema for representing *Initial* orbit state/ configuration

### DISCUSSION THREAD

**Discussion by vinay0000** [eose-tools-org/eose-api#4] [General]
The initial orbit state shall be utilized in orbit propagators. We need to support representation of the state such that it can be readily utilized by popular orbit propagators such as SGP4 (provided by Skyfield), J2 Analytical propagator (for well-maintained orbits), numerical propagators provided by GMAT.

There are multiple possible representations of the satellite orbital state. There are two parameters which are needed to  characterize the orbit are: (1) [Epoch (date)](https://github.com/eose-tools-org/eose-api/discussions/3) (2) ephemerides (e.g., Keplerian, Cartesian state vector) and its associated reference frame (e.g., J2000, ICRF).

Orbit representations in relevant software:

GMAT: https://documentation.help/GMAT/SpacecraftOrbitState.html

STK: https://help.agi.com/stk/#stk/vehSat_coordType.htm

**CCSDS Orbit Mean-Elements Message (OMM)** 

OMM maybe preferred since it also contains the ``CREATION_DATE``, which is useful in factoring in the latency of the orbital data, and hence its application in real-time applications.
The representation is in terms of mean Keplerian elements in the TEME frame at a specified epoch (UTC).

[CCSDS 502.0-B-2](https://www.nasa.gov/wp-content/uploads/2017/12/orbit_data_messages.pdf)

Example implementation with OMM (from SpaceTrack) is shown below. It has the entire OMM info (JSON format). An issue maybe that the naming convention/ type of the key/value pairs would not be in line with our convention. Alternatively, a subset of the info can be provided.
```
{"type": "omm",
	"omm":  {
        "CCSDS_OMM_VERS": "2.0",
        "COMMENT": "GENERATED VIA SPACE-TRACK.ORG API",
        "CREATION_DATE": "2024-04-15T14:46:20",
        "ORIGINATOR": "18 SPCS",
        "OBJECT_NAME": "ISS (ZARYA)",
        "OBJECT_ID": "1998-067A",
        "CENTER_NAME": "EARTH",
        "REF_FRAME": "TEME",
        "TIME_SYSTEM": "UTC",
        "MEAN_ELEMENT_THEORY": "SGP4",
        "EPOCH": "2024-04-15T11:59:15.496224",
        "MEAN_MOTION": "15.50226864",
        "ECCENTRICITY": "0.0004667",
        "INCLINATION": "51.6382",
        "RA_OF_ASC_NODE": "266.2398",
        "ARG_OF_PERICENTER": "68.1577",
        "MEAN_ANOMALY": "291.9908",
        "EPHEMERIS_TYPE": "0",
        "CLASSIFICATION_TYPE": "U",
        "NORAD_CAT_ID": "25544",
        "ELEMENT_SET_NO": "999",
        "REV_AT_EPOCH": "44882",
        "BSTAR": "0.00034083",
        "MEAN_MOTION_DOT": "0.00019203",
        "MEAN_MOTION_DDOT": "0",
        "TLE_LINE0": "0 ISS (ZARYA)",
        "TLE_LINE1": "1 25544U 98067A   24106.49948491  .00019203  00000-0  34083-3 0  9997",
        "TLE_LINE2": "2 25544  51.6382 266.2398 0004667  68.1577 291.9908 15.50226864448823",
        "SEMIMAJOR_AXIS": "6794.200",
        "PERIOD": "92.890",
        "APOAPSIS": "419.236",
        "PERIAPSIS": "412.894",
        "OBJECT_TYPE": "PAYLOAD",
        "DECAYED": "0"
    }}
```

**Two-Line Element Sets (TLEs)**
The representation is in terms of mean Keplerian elements in the TEME frame at a specified epoch.

Example implementation with TLE:
```
{	"type": "tle",
	"tle_lineo": "AQUA"
	"tle_line1": "27424U 02022A   24052.86568623  .00001525  00000-0  33557-3 0  9991"
	"tle_line2: "2 27424  98.3176   1.9284 0001998  92.8813 328.6214 14.58896689159754"
}
```
Poliastro has a nice description about the right way in which OMM and TLE data must be propagated since they contain Brouwer mean elements, which cannot be directly translated to osculating elements.

From Poliastro:

> Therefore, the correct way of using GP data is:
> 
> Propagate the TLE or OMM using the SGP4 algorithm, which produces cartesian elements (position and velocity) in the True Equator Mean Equinox (TEME) reference frame.
> 
> Convert the resulting coordinates from TEME to the desired one (GCRS/ECI, ITRS/ECEF, other) (see note below).

**Instantaneous Keplerian elements or Cartesian coordinates** (discussed [here](https://github.com/eose-tools-org/eose-api/discussions/5)) may be more amenable for numerical orbit propagators.

**Special orbits:** Orbits with certain characteristics of interest for the Earth Observation applications. The schema would have a (orbit) ``type`` specification, with required parameters.

- Circular orbits
- Sun-Synchronous orbits (SSO) with specification of local time of ascending node (LTAN)
- Daw dusk orbit: Special case of SSO with 6am LTAN
- Molniya orbit

**Constellation orbit generator:** A set of orbits corresponding to a parameterized constellation. For example the Walker Delta constellation can be implemented as follows:
```
{"@type": 'Walker Delta Constellation', 
                  "date":{},
                  "numberSatellites": 2,
                  "numberPlanes": 1,
                  "relativeSpacing": 1,
                  "alt": 500,
                  "ecc": 0.001,
                  "inc": 45,
                  "aop": 135,
}
```

@ptgrogan @ryanketzner @jsipps26 @bgorr @dselva80 

13/May edit: Updated post to focus discussion on **initial** orbit state

**Comment by vinay0000** [comment]
I have been looking into orbit propagation with SGP4 using [Skyfield](https://rhodesmill.org/skyfield/). It has been challenging to propagate with _instantaneous_ Cartesian or Keplerian orbital elements defined in the ICRF/ J2000 (ECI) frame.  This may be a required functionality for performing tradespace studies, where fictional satellite Cartesian states (orbits) shall be generated. 

The SGP4 propagator can only be initialized with _mean_ Keplerian elements in the _TEME_ frame. The challenge is how do we convert the state in the ICRF/J2000 system to mean Keplerian elements in the TEME frame.

Any ideas? This shall be relevant to the schema since we shall need to decide if we are to offer schema with orbital states defined as instantaneous elements in the J2000 frame.

Relevant links:
https://github.com/skyfielders/python-skyfield/issues/384
https://stackoverflow.com/questions/62999600/is-it-legit-to-generate-a-tle-two-line-element-file-from-instantaneous-kepler
https://rhodesmill.org/skyfield/earth-satellites.html#from-satrec

**Reply by bgorr** [reply]
Vallado has a nice discussion of mean and osculating (instantaneous) orbital elements in Chapter 9 of Fundamentals of Astrodynamics and Applications. Vallado discusses two options for the transformation from osculating to mean. We can either convert at a single point in time which has a small error associated with it, or we can propagate the osculating elements and then do least squares to get the mean orbital elements. Brouwer-Lyddane theory also provides a transformation from osculating (short-period) to mean elements. I believe Brouwer-Lyddane mean elements are an available option in STK and GMAT: https://help.agi.com/stk/index.htm#stk/vehSat_meanElements.htm

Personally I don't see any reason why we should concern ourselves with osculating orbital elements. It should be safe to assume that we are operating with mean orbital elements. Unless the tradespace exploration in question relies on extremely precise orbit knowledge, I doubt that most users would care about the inaccuracy of using mean elements vs. osculating elements. It would also be simple to later add a routine to convert from osculating to mean and vice-versa, for advanced users.

**Comment by jsipps26** [comment]
I wonder what the group is thinking in terms of these two orbit state options:

1. Analytic orbit (general perturbations)
- Initial state (mean or instantaneous)
- Epoch
- Reference frame
- Propagator (Keplerian, TLE, OMM, etc.) - the propagator implies the force model
- Note: mean vs. instantaneous state and reference frame may be implied from the propagator, but it would be best if the user can just input whatever state, epoch, and reference frame, and internal functions handle the conversions given the propagator.

2. Numerical orbit (special perturbations)
- Specify ephemeris file or specify time-series vectors of pos/vel
- Epoch
- Reference frame
- Metadata in file: propagator, force model, physical satellite characteristics

To me, the "analytic orbit" (or just "Orbit") makes the most sense to serve as a pipeline to communicate between different coverage calculators. It is unclear to me whether we can just make an optional input for the "propagator". It seems that the whole point of EOSE is that an orbit I have made in my coverage tool can be remade/used in another coverage tool without explicitly sending an entire time-series. If I want to make an orbit in TEMPO (Tool for early mission planning and observation) that can be compared to TAT-C, can I not just convert whatever my orbit ephemeris is into an SGP4 equivalent, then send it through EOSE to TAT-C?

On the flipside, while sending the entire time-series is not preferred, it might be necessary for a general numerical propagation. The issue is that force models are not well-defined... you could place in metadata that you're using "Harris-Priester" for drag but there's a host of parameters that go along with that model. For example, the drag coefficient or F10.7 may need to be specified, which may change throughout the mission lifetime. For solar radiation pressure, the types of materials on the spacecraft can matter. High-fidelity ECI to ECEF conversion can vary depending on precession and nutation models/conventions, causing variations in earth-fixed perturbations. For complex force models, the input parameters can be immense. It may be better to make a "numerical orbit" class which is effectively an I/O operator that handles reading/writing in a familiar format. Alternatively, the "numerical orbit" class could just hold a time-series vector of positions and velocities, the epoch, reference frame, and a string of metadata. From the coverage calculation perspective, we only need cartesian state and velocity, the reference frame, and the epoch. The metadata is just there to interpret outputs. 

**Question: Is the goal of EOSE to set standards in coverage analysis, or to offer a library of transformations? For example, if a user makes their own orbit from Skyfield, are they supposed to be able to compute coverage in TAT-C with that orbit?**

**Comment by ptgrogan** [comment]
I have an [example TAT-C implementation ](https://github.com/eose-tools-org/eose-api/blob/tatc-orbit-example/tatc-orbit-example.ipynb) with both initial orbit state and instantaneous orbit states. Initial orbit state is based on OMM. Instantaneous orbit states are Cartesian position/velocity using the International Celestial Reference System (ICRS) frame.

Here is an example JSON-formatted request and response snippet:

Inputs (PropagationRequest):
```json
{
  "orbit":{
    "object_name":"ISS (ZARYA)",
    "object_id":"1998-067A",
    "epoch":"2024-06-07T09:53:34.728000",
    "mean_motion":15.50975122,
    "eccentricity":0.0005669,
    "inclination":51.6419,
    "ra_of_asc_node":3.7199,
    "arg_of_pericenter":284.672,
    "mean_anomaly":139.0837,
    "ephemeris_type":0,
    "classification_type":"U",
    "norad_cat_id":25544,
    "element_set_no":999,
    "rev_at_epoch":45703,
    "bstar":0.00033759,
    "mean_motion_dot":0.00019541,
    "mean_motion_ddot":0.0
  },
  "start":"2024-01-01T00:00:00Z",
  "duration":"P1D",
  "time_step":"PT1M"
}
```

Outputs (PropagationResponse):
```json
{
  "records":[
    {
      "time":"2024-01-01T00:00:00Z",
      "position":[-4499686.1202688,-232613.21550674032,5091657.311614182],
      "velocity":[-1677.6634157866658,-7252.659671818871,-1804.8077999287968],
      "frame": "ICRS"
    },
    ...
  ]
}
```
